library/core/src/alloc/global.rs

```diff
@@ -20,24 +20,51 @@ use crate::ptr;
 ///
 /// # Example
 ///
-/// ```no_run
-/// use std::alloc::{GlobalAlloc, Layout, alloc};
-/// use std::ptr::null_mut;
-///
-/// struct MyAllocator;
+/// ```
+/// use std::alloc::{GlobalAlloc, Layout};
+/// use std::cell::UnsafeCell;
+/// use std::process::abort;
+/// use std::sync::atomic::{Ordering, AtomicUsize};
 ///
-/// unsafe impl GlobalAlloc for MyAllocator {
-///     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 { null_mut() }
-///     unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+/// const ARENA: usize = 100 * 1024;
+/// struct SimpleAllocator {
+///     arena: UnsafeCell<[u8; ARENA]>,
+///     counter: AtomicUsize,
 /// }
 ///
 /// #[global_allocator]
-/// static A: MyAllocator = MyAllocator;
+/// static ALLOCATOR: SimpleAllocator = SimpleAllocator {
+///     arena: UnsafeCell::new([0x55; ARENA]),
+///     counter: AtomicUsize::new(0),
+/// };
 ///
-/// fn main() {
-///     unsafe {
-///         assert!(alloc(Layout::new::<u32>()).is_null())
+/// unsafe impl Sync for SimpleAllocator { }
+///
+/// unsafe impl GlobalAlloc for SimpleAllocator {
+///     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+///         let align_mask = layout.align() - 1;
+///         let size = layout.size();
+///         let mut result = 0;
+///         let mut new = 0;
+///         self.counter.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |mut counter| {
+///             counter += align_mask;
+///             counter &= !align_mask;
+///             result = counter;
+///             counter += size;
+///             new = counter;
+///             Some(counter)
+///         }).unwrap();
+///         if new > ARENA { abort(); }
```

Comment:
The allocator returns null on allocation failure, it does not abort

Reply:
> The allocator returns null on allocation failure, it does not abort

Ideally, yes.  But with the current structure of the critical section (no conditional other than the one inherent in `fetch_update`, the counter would eventually overflow after many many repeated allocations.

Although this is only a demo, I think your implication, that this ought to be addressed, is right, because it's supposed to be demoing the `GlobalAlloc` trait.  I will improve this.

Reply:
I would make it limited (in features), complete and correct. Use a Mutex to simplify the code for the reader. Compute the needed bump for the allocation, if there is no more space, return null, otherwise "commit" to it and update and release the lock. Limited could mean that it only supports one alignment (everything is 16-aligned and larger aligns fail to allocate?) or other realistic restriction, just for this example. just to make the example easier to read and write.

Reply:
I like the suggestion of trying to use a Mutex.  Unfortunately they can't be constructed in const context (AIUI because of Windows).  I think `fetch_update` is a good primitive to use but I take your point that I could just bung everything relevant into the critical section which would simplify matters considerably.

The alignment comes out much nicer when counting down so it may turn out to be shorter to keep the alignment support than special case it.